### PR TITLE
Grapple

### LINKED ISSUE #201: Grapple Hook Pull

### PR DESCRIPTION
**Pull request by EdenRedd**
I made a grapple with the settings that I thought were okay, to change the settings you have to go to the grapplescript to the part where it says grapple settings,

### CODE REVIEW

File: Assets/Scripts/GrappScript.cs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+public class GrappScript : ProjectileWeapon
+{
+    SpringJoint2D joint;
+    Vector2 grapplePoint;
+    public LineRenderer lr;
+    
+    override
+    public void Fire(Vector2 direction)
+    {
+        lr = GameObject.FindGameObjectWithTag("Player").GetComponent<LineRenderer>();
```
Comment (BUNN1E5) [inline Assets/Scripts/GrappScript.cs]: Is there going to be a "Player" object that has a line renderer?

File: Assets/Scripts/GrappScript.cs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            PlatformerController.instance.StartCoroutine(DrawRope());
+        }
+
+    }
+
+    IEnumerator DrawRope()
+    {
+        while (joint)
+        {
+            lr.SetPosition(0, GameObject.FindGameObjectWithTag("Player").transform.position);
```
Comment (BUNN1E5) [inline Assets/Scripts/GrappScript.cs]: If you can avoid it, please don't use `FindGameObjectWithTag`, instead use `PlatformerController.instance.transform.position`

File: Assets/Scripts/GrappScript.cs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        Debug.DrawRay(playerPos, direction, Color.red, 10);
+        if (hit.collider != null)
+        {
+  
+
+            joint = PlatformerController.instance.gameObject.AddComponent<SpringJoint2D>();
+            Debug.Log("hit something");
+            //joint = GameObject.FindGameObjectWithTag("Player").AddComponent<SpringJoint2D>();    for testing while i didnt have PlatformerController on
+
+            grapplePoint = hit.collider.ClosestPoint(hit.point);
```
Comment (BUNN1E5) [inline Assets/Scripts/GrappScript.cs]: So what happens when your grapple hits an entity? Specifically an Enemy?

File: Assets/Scripts/GrappScript.cs (hunk 1 above)
Comment (BUNN1E5) [reply Assets/Scripts/GrappScript.cs]: Maybe just create a new one when needed? The cost isn't that much and it keeps the player cleaner.

File: Assets/Scripts/GrappScript.cs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        }
+
+    }
+
+    IEnumerator DrawRope()
+    {
+        while (joint)
+        {
+            lr.SetPosition(0, GameObject.FindGameObjectWithTag("Player").transform.position);
+            lr.SetPosition(1, grapplePoint);
```
Comment (BUNN1E5) [inline Assets/Scripts/GrappScript.cs]: This could be `lr.SetPosition(1, joint.connectedAnchor);`

File: Assets/Scripts/GrappScript.cs [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+            joint = PlatformerController.instance.gameObject.AddComponent<SpringJoint2D>();
+            Debug.Log("hit something");
+            //joint = GameObject.FindGameObjectWithTag("Player").AddComponent<SpringJoint2D>();    for testing while i didnt have PlatformerController on
+
+            grapplePoint = hit.collider.ClosestPoint(hit.point);
+
+            joint.enableCollision = true;
+            joint.autoConfigureConnectedAnchor = false;                         //grapple settings
+            joint.connectedAnchor = grapplePoint;
```
Comment (BUNN1E5) [inline Assets/Scripts/GrappScript.cs]: why even have `grapplePoint` of you set it to `joint.connectedAnchor` immediately anyways? Why not just use `joint.connectedAnchor = hit.collider.ClosestPoint(hit.point);` and then reference `joint.connectedAnchor` when wanting to get the grapplePoint?

**Review by BUNN1E5 — CHANGES_REQUESTED** [review]

File: Assets/Scripts/GrappScript.cs [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+public class GrappScript : ProjectileWeapon
+{
+    SpringJoint2D joint;
+    Vector2 grapplePoint;
+    public LineRenderer lr;
+    
+    override
+    public void Fire(Vector2 direction)
+    {
+        lr = GameObject.FindGameObjectWithTag("Player").AddComponent<LineRenderer>();
```
Comment (BUNN1E5) [inline Assets/Scripts/GrappScript.cs]: Please change with `PlatformerController.instance.AddComponent<LineRenderer>()`

File: Assets/Scripts/GrappScript.cs [hunk 7]
```diff
@@ -0,0 +1,76 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+[CreateAssetMenu(fileName = "GrappScript", menuName = "LTG5/Weapons/GrappleHook", order = 0)]
+public class GrappScript : ProjectileWeapon
+{
+    SpringJoint2D joint;
+    Vector2 grapplePoint;
```
Comment (BUNN1E5) [inline Assets/Scripts/GrappScript.cs]: Remove unused

**Review by BUNN1E5 — CHANGES_REQUESTED** [review]

File: Assets/Scripts/GrappScript.cs (hunk 3 above)
Comment (EdenRedd) [reply Assets/Scripts/GrappScript.cs]: When hitting an enemy it should bring the player slightly closer to the enemy

File: Assets/Scripts/GrappScript.cs [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+        Debug.DrawRay(playerPos, direction, Color.red, 10);
+        if (hit.collider != null)
+        {
+  
+
+            joint = PlatformerController.instance.gameObject.AddComponent<SpringJoint2D>();
+  
+
+             ;
```
Comment (BUNN1E5) [inline Assets/Scripts/GrappScript.cs:46]: Hey I think you lost a semicolon lol

**Review by BUNN1E5 — APPROVED** [review]
Other than that it looks and works good

**Pull request merged**